### ISSUE TITLE
MySQL8 throws java.sql.SQLInvalidAuthorizationSpecException

### ISSUE THREAD

**Issue by gedOHub** [gitbucket/gitbucket#2201]
Hi,

I'm unable to change H2 with MySQL (8.0.13).

What I have done:
 * I have created database gitbucket.
 * I have created user gitbucket with a password.
 * I have changed `database.conf` according to `https://github.com/gitbucket/gitbucket/wiki/External-database-configuration`

Log full of these errors:
```
...
Nov 25 11:29:06 pc java: at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
Nov 25 11:29:06 pc java: at JettyLauncher.main(JettyLauncher.java:127)
Nov 25 11:29:06 pc java: Caused by: java.sql.SQLException: Access denied for user 'gitbucket'@'localhost' (using password: NO)
Nov 25 11:29:06 pc java: Current charset is UTF-8. If password has been set using other charset, consider using option 'passwordCharacterEncoding'
...
Nov 25 11:29:06 pc java: at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
Nov 25 11:29:06 pc java: at JettyLauncher.main(JettyLauncher.java:127)
Nov 25 11:29:06 pc java: Caused by:
Nov 25 11:29:06 pc java: com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: Access denied for user 'gitbucket'@'localhost' (using password: NO)
Nov 25 11:29:06 pc java: at com.zaxxer.hikari.pool.HikariPool.throwPoolInitializationException(HikariPool.java:568)
...
Nov 25 11:29:06 pc java: at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
Nov 25 11:29:06 pc java: at JettyLauncher.main(JettyLauncher.java:127)
Nov 25 11:29:06 pc java: Caused by:
Nov 25 11:29:06 pc java: java.sql.SQLInvalidAuthorizationSpecException: Access denied for user 'gitbucket'@'localhost' (using password: NO)
Nov 25 11:29:06 pc java: at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.get(ExceptionMapper.java:173)
...
Nov 25 11:29:06 pc java: at JettyLauncher.main(JettyLauncher.java:127)
Nov 25 11:29:06 pc java: Caused by:
Nov 25 11:29:06 pc java: java.sql.SQLException: Access denied for user 'gitbucket'@'localhost' (using password: NO)|Current charset is UTF-8. If password has been set using other charset, consider using option 'passwordCharacterEncoding'
Nov 25 11:29:06 pc java: at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.authentication(AbstractConnectProtocol.java:854)
Nov 25 11:29:06 pc java: at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.handleConnectionPhases(AbstractConnectProtocol.java:777)
...
```

While researching also tried to define all settings in URL by this format:
```
jdbc:mysql://localhost/test?user=minty&password=greatsqldb
```

**Deployment mode**: 
- *Httpd*
- *Gitbucket as systemctl service*

**Problem description**:
- *Gitbucket for some reason don't read or use password configurated at database.conf file*

**Comment by takezoe** [comment]
Accoriding to the documentation, specify username and password as another properties as:
```
db {
  url = "jdbc:mysql://localhost/gitbucket?useUnicode=true&characterEncoding=utf8"
  user = "testuser"
  password = "testpassword"
}
```

**Comment by KJunya** [comment]
I'm having the same problem.
I've deployed gitbucket using tomcat, but whenever I edit database.conf to use mysql instead of h2, I get a 404 error on url `http://192.168.1.2:8080/gitbucket/`
The configuration of my system are as follows, but what am I missing?

- os
```
CentOS Linux release 7.6.1810 (Core)
```
- java
```
openjdk 11.0.2 2019-01-15 LTS
OpenJDK Runtime Environment 18.9 (build 11.0.2+7-LTS)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.2+7-LTS, mixed mode, sharing)
```
- tomcat
```
Apache Tomcat/9.0.16
```
- mysql
```
mysql  Ver 8.0.15 for Linux on x86_64 (MySQL Community Server - GPL)
```
- environments
```
export JAVA_HOME=/usr/lib/jvm/java-11-openjdk-11.0.1.13-3.el7_6.x86_64
export PATH=$JAVA_HOME/bin:$PATH
export CATALINA_HOME=/opt/tomcat
export GITBUCKET_HOME=/var/gitbucket
```
- /var/gitbucket/database.conf
```
db {
  url = "jdbc:mysql://localhost/gitbucket?useUnicode=true&characterEncoding=utf8"
  user = "gitbucket"
  password = "pas!w_rd" 
}
```
\
I've setup the mysql db with the following commands
```
CREATE DATABASE gitbucket CHARACTER SET utf8 COLLATE utf8_general_cli;
CREATE USER gitbucket IDENTIFIED BY 'pas!w_rd';
GRANT ALL ON gitbucket.* TO gitbucket@localhost;
FLUSH PRIVILEGES;
```

\
This is my log at /opt/tomcat/logs/localhost.2019-03-10.log
```
10-Mar-2019 21:12:12.350 INFO [Thread-4] org.apache.catalina.core.ApplicationContext.log SessionListener: contextDestroyed()
10-Mar-2019 21:12:12.351 INFO [Thread-4] org.apache.catalina.core.ApplicationContext.log ContextListener: contextDestroyed()
10-Mar-2019 21:12:49.653 SEVERE [main] org.apache.catalina.core.StandardContext.listenerStart Exception sending context initialized event to listener instance of class [gitbucket.core.servlet.InitializeLi
stener]
 java.lang.ExceptionInInitializerError
	at gitbucket.core.servlet.InitializeListener.contextInitialized(InitializeListener.scala:59)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4668)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5136)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:713)
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:690)
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:695)
	at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:978)
	at org.apache.catalina.startup.HostConfig$DeployWar.run(HostConfig.java:1850)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)
	at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:118)
	at org.apache.catalina.startup.HostConfig.deployWARs(HostConfig.java:773)
	at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:427)
	at org.apache.catalina.startup.HostConfig.start(HostConfig.java:1577)
	at org.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:309)
	at org.apache.catalina.util.LifecycleBase.fireLifecycleEvent(LifecycleBase.java:123)
	at org.apache.catalina.util.LifecycleBase.setStateInternal(LifecycleBase.java:424)
	at org.apache.catalina.util.LifecycleBase.setState(LifecycleBase.java:367)
	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:929)
	at org.apache.catalina.core.StandardHost.startInternal(StandardHost.java:831)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1377)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1367)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)
	at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:140)
	at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:902)
	at org.apache.catalina.core.StandardEngine.startInternal(StandardEngine.java:262)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.StandardService.startInternal(StandardService.java:423)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.StandardServer.startInternal(StandardServer.java:928)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.startup.Catalina.start(Catalina.java:638)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.catalina.startup.Bootstrap.start(Bootstrap.java:350)
	at org.apache.catalina.startup.Bootstrap.main(Bootstrap.java:492)
Caused by: com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: Access denied for user 'gitbucket'@'localhost' (using password: NO)
	at com.zaxxer.hikari.pool.HikariPool.throwPoolInitializationException(HikariPool.java:576)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:562)
	at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:115)
	at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:81)
	at gitbucket.core.servlet.Database$.<init>(TransactionFilter.scala:65)
	at gitbucket.core.servlet.Database$.<clinit>(TransactionFilter.scala)
	... 42 more
Caused by: java.sql.SQLInvalidAuthorizationSpecException: Access denied for user 'gitbucket'@'localhost' (using password: NO)
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.get(ExceptionMapper.java:232)
	at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.getException(ExceptionMapper.java:165)
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.connectWithoutProxy(AbstractConnectProtocol.java:1199)
	at org.mariadb.jdbc.internal.util.Utils.retrieveProxy(Utils.java:560)
	at org.mariadb.jdbc.MariaDbConnection.newConnection(MariaDbConnection.java:174)
	at org.mariadb.jdbc.Driver.connect(Driver.java:92)
	at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:136)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:369)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:198)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:467)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:541)
	... 46 more
Caused by: java.sql.SQLException: Access denied for user 'gitbucket'@'localhost' (using password: NO)
Current charset is UTF-8. If password has been set using other charset, consider using option 'passwordCharacterEncoding'
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.authentication(AbstractConnectProtocol.java:931)
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.handleConnectionPhases(AbstractConnectProtocol.java:850)
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.connect(AbstractConnectProtocol.java:507)
	at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.connectWithoutProxy(AbstractConnectProtocol.java:1195)
	... 54 more
```

**Comment by takezoe** [comment]
Have you tried `passwordCharacterEncoding` according to the error message?

**Comment by KJunya** [comment]
Hi, thanks for the swift response.
Unfortunately, I couldn't resolve anything by setting `passwordCharcterEncoding`.
But, then again, am I even doing this right?
```
db {
  url = "jdbc:mysql://localhost/gitbucket?useUnicode=true&characterEncoding=utf8&passwordCharacterEncoding=utf8"
  user = "gitbucket"
  password = "pas!w_rd"
}
```

What concerns me the most is that the log shows `(using password: NO)`, even though I clearly have set it in config.\\
Why is it not using password?

**Comment by takezoe** [comment]
Hmm... Can you access to the database using any JDBC client tool like [DBeaver](https://dbeaver.io/) or [SquirrelSQL](http://squirrel-sql.sourceforge.net/) with above username and password?

**Comment by gedOHub** [comment]
I was able to resolve this issue only when changed MySQL server to 5.7 instead of 8.0.

My situation let me do that.

**Comment by KJunya** [comment]
Sorry for the late reply.

Although it was from my remote machine as I do not have windowing system installed in the server, I was able to connect to mysql db using DBeaver with user gitbucket and gitbucket database.

I've also tried postgresql 9.2.24 and 11.2 with `url = "jdbc:postgresql://localhost/gitbucket"` which both were able to display the gitbucket page.

I haven't tried mysql 5.7 yet, however, even if it does resolve the issue I still don't want to hassle to run different versions of mysql.

So, in conclusion in the mean time I'll use postgresql 11 instead of mysql 8 for gitbucket.

Thanks for your quick response and help.

**Comment by nao-sgm** [comment]
We report the same problem as it occurred.

・gitbucket ver 4.31.1

・CentOS Linux release 7.5.1804 (Core)
・mysql  Ver 8.0.15 for Linux on x86_64 (MySQL Community Server - GPL)

・Java
openjdk version "1.8.0_201"
OpenJDK Runtime Environment (build 1.8.0_201-b09)
OpenJDK 64-Bit Server VM (build 25.201-b09, mixed mode)

・Tomcat
Server version: Apache Tomcat/8.5.38

This is my log
----------------------
java.sql.SQLInvalidAuthorizationSpecException: Access denied for user 'gitbucket'@'localhost' (using password: NO)
        at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.get(ExceptionMapper.java:232)
        at org.mariadb.jdbc.internal.util.exceptions.ExceptionMapper.getException(ExceptionMapper.java:165)
        at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.connectWithoutProxy(AbstractConnectProtocol.java:1199)
        at org.mariadb.jdbc.internal.util.Utils.retrieveProxy(Utils.java:560)
        at org.mariadb.jdbc.MariaDbConnection.newConnection(MariaDbConnection.java:174)
        at org.mariadb.jdbc.Driver.connect(Driver.java:92)
        at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:136)
        at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:369)
        at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:198)
        at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:467)
        at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:541)
        at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:115)
        at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:81)
        at gitbucket.core.servlet.Database$.<init>(TransactionFilter.scala:65)
        at gitbucket.core.servlet.Database$.<clinit>(TransactionFilter.scala)
        at gitbucket.core.servlet.InitializeListener.contextInitialized(InitializeListener.scala:59)
        at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4817)
        at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5283)
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
        at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:754)
        at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:730)
        at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:734)
        at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:985)
        at org.apache.catalina.startup.HostConfig$DeployWar.run(HostConfig.java:1857)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.sql.SQLException: Access denied for user 'gitbucket'@'localhost' (using password: NO)
Current charset is UTF-8. If password has been set using other charset, consider using option 'passwordCharacterEncoding'
        at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.authentication(AbstractConnectProtocol.java:931)
        at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.handleConnectionPhases(AbstractConnectProtocol.java:850)
        at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.connect(AbstractConnectProtocol.java:507)
        at org.mariadb.jdbc.internal.protocol.AbstractConnectProtocol.connectWithoutProxy(AbstractConnectProtocol.java:1195)
        ... 26 common frames omitted

**Comment by shimamoto** [comment]
The root cause is probably the following reason:
https://stackoverflow.com/questions/54188309/mariadb-connection-client-access-denied-for-user-using-password-no-on-mysql

**Comment by nao-sgm** [comment]
Thanks for the answer. It was very helpful.
It was a problem on MySql side. I'm sorry.

This solved it.
```
ALTER  USER 'gitbucket'@'localhost' IDENTIFIED  WITH  mysql_native_password  BY 'PASSWORD';

mysql> SELECT  user, host, plugin  FROM  mysql.user;
+------------------+-----------+-----------------------+
| user             | host      | plugin                |
+------------------+-----------+-----------------------+
| gitbucket        | localhost | mysql_native_password |
| mysql.infoschema | localhost | caching_sha2_password |
| mysql.session    | localhost | caching_sha2_password |
| mysql.sys        | localhost | caching_sha2_password |
| root             | localhost | mysql_native_password |
+------------------+-----------+-----------------------+
```

**Comment by takezoe** [comment]
Great. Thanks @shimamoto for your help. I updated the documentation:
https://github.com/gitbucket/gitbucket/wiki/External-database-configuration#mysql-57-or-higher

I think changing the authentication method in MySQL side is a temporary solution until MariaDB connector supports it. We need to watch https://jira.mariadb.org/browse/CONJ-663 and upgrade MariaDB connector. I would like to keep this issue open until that.

**Comment by takezoe** [comment]
This issue has been fixed in GitBucket 4.33.0 by https://github.com/gitbucket/gitbucket/pull/2409.
